toolchain/check/import_ref.cpp

```diff
@@ -67,21 +68,36 @@ class ImportRefResolver {
   auto Resolve(SemIR::InstId inst_id) -> SemIR::ConstantId {
     work_stack_.push_back({inst_id});
     while (!work_stack_.empty()) {
+      auto initial_work = work_stack_.size();
       auto work = work_stack_.back();
       CARBON_CHECK(work.inst_id.is_valid());
 
       // Double-check that the constant still doesn't have a calculated value.
       // This should typically be checked before adding it, but a given
       // instruction may be added multiple times before its constant is
       // evaluated.
-      if (!work.made_forward_decl &&
-          import_ir_constant_values_.Get(work.inst_id).is_valid()) {
+      //
+      // If a retry was requested, then rerun this work item even if it already
+      // has a constant value from the previous run. It may not have been fully
+      // imported, even though it was imported enough to produce a constant.
```

Comment:
Suggestion, you might collapse this text together with something like:

We can add an instruction to the stack multiple times. If this occurs, we'll already have a constant for it, and should not call TryResolveInst. However, we may also have a constant when an instruction is being retried, in which case we do need to call TryResolveInst.

Reply:
Made the comments in this function more directly call back to the class comment.

I think it's useful to have the big comment on the class because it's useful context for the `TryResolve*` functions too.